Add current vs. legacy differences page for upgraders

nfagerlund opened the pull request.

armchairlinguist commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+### Workspaces Replace Environments
+
+Environments are now called [workspaces](../workspaces/index.html). Our goal in this change was to make it easier to on-board new Terraform users in your organization by avoiding an overloaded and ambiguous term.
```
I wonder if we could be stronger and just say "Using a dedicated term makes it easier to onboard..."

armon replied:
Instead of the second sentence, maybe "Our goal in this change was to make the terminology used by Terraform Enterprise consistent with Terraform Open Source."

mitchellh replied:
I would also add something like: "Beyond the name change, the functionality of workspaces should be the same as what was previously known as 'environments.'"

armchairlinguist commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+Legacy Terraform Enterprise included an event log that logged details about run events. For the new Terraform Enterprise, we replaced it with a [comprehensive audit logging system][audit] that also logs all changes to organization and workspace settings.
+
+The new audit log is available on [private installs](../private/index.html) of Terraform Enterprise.
```
It's actually running on SaaS TFE as well, though I'm not sure whether we want to mention that.

nfagerlund replied:
I'd go with no! It's running... but it's not really _available_ to me as a user.

mitchellh replied:
Agreed. 

I wonder if its work linking to any docs (maybe `../private` has it) related to the new audit log, or clarifying: "other than the name change, the audit log contains the same information blah blah"

nfagerlund replied:
The "comprehensive audit logging system" link goes to the audit log docs, but the link should probably be down in the second paragraph instead. 👍

armchairlinguist commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+Legacy Terraform Enterprise could grant access permissions to individual users; the new Terraform Enterprise handles all workspace permissions via [teams of users](../users-teams-organizations/teams.html).
+
+We removed individual access because it eventually makes your workspace settings a mess. Team-based access is easier to update and govern.
```
I might say something like "complicated and difficult to manage" vs a mess, although the unvarnished version is nice too. :)

mitchellh replied:
I would also add that you can mimic individual access by creating a team with a single individual. Unless I'm mistaken.

armchairlinguist commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+We removed this feature because it worked against Terraform Enterprise's goal of collaborative and predictable infrastructure. Not only did it make runs inconsistent and less reliable, it also expected that users would be queueing most runs manually in the UI, when we've found that people want most runs to happen in response to changes in a VCS repo or other automatic triggers.
+
+If you used personal variables to avoid sharing credentials, you can mark workspace variables as sensitive.
```
Should there be a link here to sensitive variables info?

armchairlinguist commented:
As someone who's been guiding people through the transition process, I wonder if this page should make more mention of the primacy of the VCS-based workflow for the UI. That's been a big surprise for many customers.

nfagerlund commented:
@armchairlinguist Oh, word? Yeah, if that's been surprising people we should totally say something about it here.

KFishner commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+# Differences Between Current and Legacy Terraform Enterprise
+
+Terraform Enterprise began as part of the discontinued Atlas suite, and today's version is a ground-up redesign. If you've previously used the legacy version of Terraform Enterprise, we hope you enjoy the new version's improvements.
```
and today's version is a redesign purpose-built for using Terraform in a team setting.

nfagerlund replied:
👍

mitchellh replied:
I'd add some honesty in here about the unfortunate nature that we have to remove some features. "we hope you enjoy the new version's improvements. Unfortunately, as part of this redesign, some features have been removed and others are not yet implemented."

nfagerlund replied:
👍

KFishner commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+## Permanently Removed
+
+We deliberately removed these features to streamline and improve Terraform Enterprise's core features. Some of these features worked against or distracted from what Terraform Enterprise does best, and some are good ideas that are handled better by external services.
```
Terraform Enterprise's core _workflow_

nfagerlund replied:
👍🏼

mitchellh replied:
Would recommend adding another paragraph below here again empathizing with the users who are affected by this. "We understand that removing features can be painful and disappointing for those who have built processes around them. For each of these features, we ensured only a single digit percentage of users would be affected, and have attempted to mitigate the change with guides to other systems when appropriate. If you're being majorly affected by these removals and the documentation we've created isn't adequate, please contact us and we'll work with you to help you migrate."

KFishner commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+We removed individual access because it eventually makes your workspace settings a mess. Team-based access is easier to update and govern.
+
+### Personal Environment and Organization Variables
```
@skierkowski should we move this to the "Gone but returning section" and mention variable sets?

armon replied:
We should maybe move "Organizational Variables" to Returning soon and explain that Variable Sets are more flexible and encompass the same goal.

KFishner commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+This does not mean we're dropping support for a push-based workflow — we want push to be better, and the integrated `push` command was holding us back because we built it on an outdated understanding of how Terraform Enterprise should work.
+
+TFE's API supports uploading a configuration tarball and running it, and the [API-based run workflow page](../workspaces/run-api.html) explains how to integrate it into your automated systems. We also plan to release an external tool that mimics `terraform push`'s workflow soon.
```
is the plan to update this section once the `terraform push` tool comes out this week? or are both coming out at the same time?

I would move this into the "Renamed or Moved Features" section since you get the same functionality, it's just called something new

nfagerlund replied:
Yeah, I was gonna update it. Good point about moving it, although I might delay that until we've shipped the tool.

armon replied:
I would move to "Moved" and also mention the deeper CLI integration plans.

mitchellh replied:
Agreed with armon.

armon commented on content/source/docs/enterprise/upgrade/differences.html.md:
```diff
@@ -0,0 +1,78 @@
+
+Terraform Enterprise no longer includes an integrated Packer build system or an artifact registry for built images. Different organizations have such different requirements and expectations for image builds that we believe [running Packer out of your own CI/CD system](https://www.packer.io/guides/packer-on-cicd/index.html) will usually provide a better experience.
+
+### Periodically Queued Plans
```
This feels like something we might need to add back at some point, so "Permanent Removal" seems strong. Forcing the user to have a cron hit an API is a bit kludgy.

nfagerlund commented:
Ok, I think I've incorporated everyone's feedback in this commit! I've added a link to the push script repo, since it's releasing soon; we can time the merge to go along with that. Also, @armon, is that a useful but not-too-prematurely-detailed description of the enhanced backend plans?

armon commented:
@nfagerlund I think the enhanced backend description is useful. I would also talk about Variable sets as a replacement to organizational variables coming back.

nfagerlund commented:
Cool cool. I've disentangled the issues of shared and personal variables, and put a note about variable sets in the "returning soon" section.

nfagerlund commented:
Rebased to resolve merge conflicts and make less of a mess.
